build-tools/heroku/generate-rules-highlighting.py

```diff
@@ -0,0 +1,452 @@
+        for command in keywordLangByLevel["SP_K"]: 
+            list_rules.append({
+                'regex': START_WORD + "("+ command + ')([0-9]+)' + END_WORD,
+                'token': ['text','keyword','variable'],
+                'next': 'start',
+            })
+
+        for command in keywordLangByLevel["K"]:
+            list_rules.append({
+                'regex': "(" + command + ')([0-9]+)' + END_WORD,
+                'token': ['keyword','variable'],
+                'next': 'start',
+            })
+
+
+    # Rules for commands of SP_K_SP 
+    # These are the keywords that must be "alone" so neither preceded nor followed directly by a word 
+    for command in keywordLangByLevel["SP_K_SP"]:
+        list_rules.append({
+            'regex': START_WORD + command + END_WORD,
+            'token': "keyword",
+            'next': "start", 
+        })
+    
+
+    # Rules for commands of K 
+    #  These are the keywords that are independent of the context (formerly the symbols
+    # In particular, even if they are between 2 words, the syntax highlighting will select them
+    for command in keywordLangByLevel["K"]:
+        list_rules.append({
+            'regex': command,
+            'token': "keyword",
+            'next': "start", 
+        })
+
+    # Rules for commands of SP_K 
+    #  This category of keywords allows you to have keywords that are not preced
+    # by another word, but that can be followed immediately by another word. (see the PR #2413)*/
+    for command in keywordLangByLevel["SP_K"]:
+        list_rules.append({
+            'regex': START_WORD + command,
+            'token': "keyword",
+            'next': "start", 
+        })
+
+    # Rules for commands of K_SP 
+    #  This category of keywords allows you to have keywords that can be preceded immediate
+    # by another word, but that are not followed by another word.*/
+    for command in keywordLangByLevel["K_SP"]:
+        list_rules.append({
+            'regex': command + END_WORD,
+            'token': "keyword",
+            'next': "start", 
+        })
+
+    return list_rules
+
+
+# This variable lists all the keywords in each level, i.e. everything that should be displayed in red (of type `keyword`)
+# 
+# There are several categories of keywords: 
+# - SP_K_SP
+#   These are the keywords that must be "alone" so neither preceded nor followed directly by a word 
+# 
+# - K
+#   These are the keywords that are independent of the context (formerly the symbols).
+#   In particular, even if they are between 2 words, the syntax highlighting will select them
+# 
+# - SP_K
+#   This category of keywords allows you to have keywords that are not preceded
+#   by another word, but that can be followed immediately by another word. (see the PR #2413)
+# 
+# - K_SP
+#   This category of keywords allows you to have keywords that can be preceded immediately
+#   by another word, but that are not followed by another word.
+KEYWORDS = {
```

Comment:
I see you adding abstraction and convience here. Nice. I don't understand why parts of the rules seem to be autogenerated and parts of it seem to be manually put in?

Reply:
Lots of repetition in here as well. That's generally not great for maintainability.

Reply:
ok (maybe make an issues for this assigned to yourself, just so we don't forget?)

Reply:
This exit already exists! (I had the wrong link)